panels/updates/cc-updates-panel.c

```diff
@@ -0,0 +1,688 @@
+
+  g_cancellable_cancel (self->cancellable);
+  g_clear_object (&self->cancellable);
+  g_clear_object (&self->current_device);
+  g_clear_object (&self->nm_client);
+
+  G_OBJECT_CLASS (cc_updates_panel_parent_class)->finalize (object);
+}
+
+static void
+cc_updates_panel_set_property (GObject      *object,
+                               guint         prop_id,
+                               const GValue *value,
+                               GParamSpec   *pspec)
+{
+  switch (prop_id)
+    {
+    case PROP_PARAMETERS:
+      /* Nothing to do */
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+cc_updates_panel_class_init (CcUpdatesPanelClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+  GtkWidgetClass *widget_class = GTK_WIDGET_CLASS (klass);
+
+  object_class->dispose = cc_updates_panel_dispose;
+  object_class->finalize = cc_updates_panel_finalize;
+  object_class->set_property = cc_updates_panel_set_property;
+
+  g_object_class_override_property (object_class, PROP_PARAMETERS, "parameters");
+
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/control-center/updates/cc-updates-panel.ui");
+
+  gtk_widget_class_bind_template_child (widget_class, CcUpdatesPanel, automatic_updates_container);
+  gtk_widget_class_bind_template_child (widget_class, CcUpdatesPanel, automatic_updates_switch);
+  gtk_widget_class_bind_template_child (widget_class, CcUpdatesPanel, metered_data_label);
+  gtk_widget_class_bind_template_child (widget_class, CcUpdatesPanel, network_name_label);
+  gtk_widget_class_bind_template_child (widget_class, CcUpdatesPanel, network_settings_label);
+  gtk_widget_class_bind_template_child (widget_class, CcUpdatesPanel, network_status_icon);
+  gtk_widget_class_bind_template_child (widget_class, CcUpdatesPanel, scheduled_updates_switch);
+  gtk_widget_class_bind_template_child (widget_class, CcUpdatesPanel, tariff_editor);
+
+  gtk_widget_class_bind_template_callback (widget_class, on_automatic_updates_switch_changed_cb);
+  gtk_widget_class_bind_template_callback (widget_class, on_change_network_link_activated_cb);
+  gtk_widget_class_bind_template_callback (widget_class, on_scheduled_updates_switch_changed_cb);
+  gtk_widget_class_bind_template_callback (widget_class, on_tariff_changed_cb);
+
+  g_type_ensure (CC_TYPE_TARIFF_EDITOR);
+}
+
+static void
+cc_updates_panel_init (CcUpdatesPanel *self)
+{
+  g_autofree gchar *settings_text = NULL;
+
+  g_resources_register (cc_updates_get_resource ());
+
+  gtk_widget_init_template (GTK_WIDGET (self));
+
+  /* "Change Network Settings" label */
+  settings_text = g_strdup_printf ("<a href=\"\">%s</a>", _("Change Network Settings…"));
+  gtk_label_set_markup (GTK_LABEL (self->network_settings_label), settings_text);
+
+  /* Load any saved tariff */
+  load_tariff (self);
+
+  /* Setup network manager */
+  self->nm_client = nm_client_new (NULL, NULL);
```

Comment:
This should be async. It creates a D-Bus proxy to connect to NetworkManager.

Reply:
It is common (but not good) practice in Settings to load this synchronously. I am going to rework that upstream for 3.28, and since by then we'd have to rework this initialization code anyway, I concluded it's not worth the hassle.

Reply:
OK. Please leave a `/* FIXME: Rework this to be async. This will be done at the same time as reworking all the other panels to be async. */` comment there to make that obvious.